src/libcrun/linux.c

```diff
@@ -1659,6 +1675,34 @@ do_notify_socket (libcrun_container_t *container, const char *rootfs, libcrun_er
   if (ret < 0)
     return crun_make_error (err, errno, "mkdir `%s`", host_notify_socket_path);
 
+  if (get_private_data (container)->unshare_flags & CLONE_NEWUSER)
+    {
+      get_root_in_the_userns_for_cgroups (container->container_def, 0, 0, &container_root_uid, &container_root_gid);
+      if (container_root_uid != ((uid_t) -1) && container_root_gid != ((gid_t) -1))
+        {
+          ret = chown (host_notify_socket_path, container_root_uid, container_root_gid);
+          if (UNLIKELY (ret < 0))
+            return crun_make_error (err, errno, "chown %d:%d `%s`", container_root_uid, container_root_gid,
+                                    host_notify_socket_path);
+        }
+    }
+
+#  if HAVE_OPEN_TREE && HAVE_AT_RECURSIVE
+  notify_socket_tree_fd = syscall_open_tree (AT_FDCWD, host_notify_socket_path, OPEN_TREE_CLONE | AT_RECURSIVE);
```

Comment:
`ENOSYS` should be handled as if HAVE_OPEN_TREE && HAVE_AT_RECURSIVE is not defined.

I'd suggest to rewrite syscall_open_tree to be always defined and return ENOSYS if HAVE_OPEN_TREE && HAVE_AT_RECURSIVE are not, so we need to handle only one fallback.